### PR TITLE
Initial code release

### PR DESCRIPTION
**Pull request by hpractv**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,79 +1,132 @@
...
+
+#### `key-name`
+
+The resulting environment variable name. It does accept a wide variety of name convention formats including spaces, dashes, periods, and underscores. While it's been tested and found to be fairly flexible, it is possible to make environment variable names that aren't usable by one or more of the target action runner operating systems.
+
+**_Key names intended for use as output variables for in workflow references can only contain letters, numbers, dashes and underscores._** Any other punctuation or space characters cannot processed and will cause errors when attempting to reference them later in a workflow.
+
+#### `scope-array`
+
+A space delimited array of scope strings that the environment variable could be valid for.  The filter for which keys/values will be select from is the `scope` action input.
```
Comment (JosephDSchwartz) [inline README.md:100]: The second sentence reads a bit strangely to me. This was my stab at trying to make it better, but feel free to do something else.
```suggestion
A space delimited array of scope strings that the environment variable could be valid for.  The filter by which keys/values will be selected is the `scope` action input.
```

File: src/action_library.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  keyName = key => key.split('@')[0];
+
+  keyIsPartOfScope = (scope, key) => {
+    let scopes = key.split('@')[1].split(' ');
+    return scopes.includes(scope);
+  };
+
+  buildEnvironmentDictionary = (scope, input, environment) => {
+    let environmentDictionary = {};
+    let inputScoped = this.filterObjectByKeyPart(input, k => this.keyIsPartOfScope(scope, k), this.keyName);
```
Comment (JosephDSchwartz) [inline src/action_library.js:39]: I'm curious how this is working. `filterObjectByKeyPart`'s third parameter should be a function based on how it's used, but `this.keyName` is a string...am I missing something super obvious?

File: src/action_library.js [hunk 3]
```diff
@@ -0,0 +1,67 @@
+const core = require('@actions/core');
+const fs = require('fs');
+const { env } = require('process');
+const yaml = require('yaml');
+
+class action_library {
+  constructor() {
+    this.currentEnvironmentPath = process.env['GITHUB_ENV'];
```
Comment (JosephDSchwartz) [inline src/action_library.js:8]: I don't see any usages for this, is it necessary?

**Review by JosephDSchwartz — DISMISSED** [review]

**Review by danielle-casella-adams — CHANGES_REQUESTED** [review]
It looks like the increment-version-on-merge.yml file was deleted....was that intentional?

I also haven't looked at the code, just the other files.

File: README.md [hunk 4]
```diff
@@ -1,79 +1,132 @@
...
 
 ## Inputs
-| Parameter | Is Required | Default | Description           |
-| --------- | ----------- | ------- | --------------------- |
-| `input-1` | true        |         | Description goes here |
-| `input-2` | false       |         | Description goes here |
+
+| Parameter                 | Is Required | Description                                                                                                                               |
+| ------------------------- | ----------- | ----------------------------------------------------------------------------------------------------------------------------------------- |
+| `scope`                   | true        | The selection filter of the vars to be set                                                                                                |
```
[the remark this answers is not in the document]
Comment (hpractv) [reply README.md:22]: I'll clean this up to be more clear.

File: README.md [hunk 5]
```diff
@@ -1,79 +1,132 @@
...
+| `input-file`              | false       | A file containing possible environment variable candidates with their associated scopes                                                   |
```
[the remark this answers is not in the document]
Comment (hpractv) [reply README.md:23]: Will clarify.

File: README.md [hunk 6]
```diff
@@ -1,79 +1,132 @@
...
 
 ## Usage Examples
 
 ```yml
-# TODO: Fill in the correct usage
+...
+input:
+  environment:
+    description: The environment to deploy to.
+    required: true
```
[the remark this answers is not in the document]
Comment (hpractv) [reply README.md:37]: 👍🏽

[the remark this answers is not in the document]
Comment (hpractv) [reply README.md:37]: 👍🏽 This input is important to understanding the example as a whole, so I'll include the whole trigger fragment.

File: README.md [hunk 7]
```diff
@@ -1,79 +1,132 @@
...
-| Parameter | Is Required | Default | Description           |
-| --------- | ----------- | ------- | --------------------- |
-| `input-1` | true        |         | Description goes here |
-| `input-2` | false       |         | Description goes here |
+
+| Parameter                 | Is Required | Description                                                                                                                               |
+| ------------------------- | ----------- | ----------------------------------------------------------------------------------------------------------------------------------------- |
+| `scope`                   | true        | The selection filter of the vars to be set                                                                                                |
+| `input-file`              | false       | A file containing possible environment variable candidates with their associated scopes                                                   |
+| `create-output-variables` | false       | Create output variables (in addiction to environment variables) for use in other steps and jobs, accepts true or false, defaults to false |
```
[the remark this answers is not in the document]
Comment (hpractv) [reply README.md:25]: 👍🏽

File: README.md [hunk 8]
```diff
@@ -1,79 +1,132 @@
...
+        with:
+          scope: ${{ workflow.inputs.environment }}
+          create-output-variables: true
+        env:
+          ENVIRONMENT@dev d development: dev
+          ENVIRONMENT@qa a: qa
+          ENVIRONMENT@stage s stg: stage
+          ENVIRONMENT@prod p production: prod
+
+      - run: echo "The current environment is ${{ env.ENVIRONMENT }}.
```
[the remark this answers is not in the document]
Comment (hpractv) [reply README.md:58]: 👍🏽

File: README.md [hunk 9]
```diff
@@ -1,79 +1,132 @@
...
+
+    build:
```
[the remark this answers is not in the document]
Comment (hpractv) [reply README.md:60]: 👍🏽

File: README.md [hunk 10]
```diff
@@ -1,79 +1,132 @@
...
-          input-2: '123
+          script: |
+            console.log("Env", ${{ needs.setup.outputs.env-scope }})
+            console.log("keyName1", ${{ env.keyName1 }})
+            console.log("keyName2", ${{ env.keyName2 }})
+            console.log("inFileName1", ${{ env.inFileName1 }})
+            console.log("inFileName2", ${{ env.inFileName2 }})
+...
+```
+
```
[the remark this answers is not in the document]
Comment (hpractv) [reply README.md:85]: I can show the file that went into the example.

File: src/main.js [hunk 11]
```diff
@@ -1,3 +1,21 @@
...
+let inputYaml = library.getFileYaml(inputFilePath);
+
+let environmentYaml = library.getCurrentEnvironmentVars();
+let environmentDictionary = library.buildEnvironmentDictionary(scope, inputYaml, environmentYaml);
+
+console.log('Scoped Variables:', environmentDictionary);
+for (envVar in environmentDictionary) {
+  library.setEnvironmentVar(`${envVar}`, environmentDictionary[envVar]);
+  if (createOutputVariables) {
+    action_library.setOutputVar(`${envVar}`, environmentDictionary[envVar]);
```
Comment (danielle-casella-adams) [inline src/main.js:19]: Should this be using `library` like the others are?  And same question about using the template string...why?
```suggestion
    library.setOutputVar(`${envVar}`, environmentDictionary[envVar]);
```

File: src/main.js [hunk 12]
```diff
@@ -1,3 +1,21 @@
...
+let scope = core.getInput('scope');
+let createOutputVariables = core.getInput('create-output-variables') == 'true';
+
+let inputFilePath = core.getInput('input-file');
+let inputYaml = library.getFileYaml(inputFilePath);
+
+let environmentYaml = library.getCurrentEnvironmentVars();
+let environmentDictionary = library.buildEnvironmentDictionary(scope, inputYaml, environmentYaml);
+
+console.log('Scoped Variables:', environmentDictionary);
```
Comment (danielle-casella-adams) [inline src/main.js:15]: Is this a `console.log` because you want it to print an object nicely?

File: src/main.js [hunk 13]
```diff
@@ -1,3 +1,21 @@
...
+for (envVar in environmentDictionary) {
+  library.setEnvironmentVar(`${envVar}`, environmentDictionary[envVar]);
```
Comment (danielle-casella-adams) [inline src/main.js:17]: Can you pass this without the template string (the method that uses it puts it in a template string too)?
```suggestion
  library.setEnvironmentVar(envVar, environmentDictionary[envVar]);
```

**Review by danielle-casella-adams — CHANGES_REQUESTED** [review]

**Comment by danielle-casella-adams** [comment]
@hpractv Have you been able to test this out?

File: README.md [hunk 14]
```diff
@@ -1,79 +1,132 @@
...
+### Usage Instructions
```
[the remark this answers is not in the document]
Comment (hpractv) [reply README.md:109]: Added a secrets example section.

**Comment by hpractv** [comment]
> @hpractv Have you been able to test this out?

Yes, it works well.

File: src/action_library.js (hunk 2 above)
Comment (JosephDSchwartz) [reply src/action_library.js:38]: ...I was missing something super obvious 😆. This isn't the first time I've made a mistake similar to this either. I'm not sure what it is about javascript function properties that are one liners like this, but I space that you actually need to invoke them to get the value 🤦‍♂️

File: src/main.js (hunk 12 above)
Comment (hpractv) [reply src/main.js:15]: Yes.  I wanted this to be easily read in the logs.

File: src/main.js (hunk 13 above)
Comment (hpractv) [reply src/main.js:17]: I can take these out.  This was left over from some test I was doing with scenarios that included spaces and commas.

File: src/main.js (hunk 11 above)
Comment (hpractv) [reply src/main.js:19]: 👍🏽. This was part of what I added this morning...should have looked more closely to what I was using to make the call.

**Review by JosephDSchwartz — COMMENTED** [review]
Everything I noticed has been addressed. I'll let Danielle give her approval when she's ready and for now just dismiss my change request review.

File: README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (danielle-casella-adams) [reply README.md]: I like the change.

File: README.md [hunk 15]
```diff
@@ -1,79 +1,132 @@
...
+          ENVIRONMENT@stage s stg: stage
+          ENVIRONMENT@prod p production: prod
+
+      - run: echo "The current environment is ${{ env.ENVIRONMENT }}.
+
+    build:
+      - name: Checkout
+        uses: actions/checkout@v2
+
+      - name: Build Workflow Environment Variables
```
[the remark this answers is not in the document]
Comment (danielle-casella-adams) [reply README.md:79]: I like this comment...I think it really helps to point out the difference here between the one in the first job.

**Review by danielle-casella-adams — APPROVED** [review]

**Pull request merged**